File: expo/utils/fsrs.ts
import { GapItem, FsrsState } from '@/types';

export type FsrsGrade = 1 | 2 | 3 | 4;

const W: number[] = [
  0.4, 0.6, 2.4, 5.8, 4.93, 0.94, 0.86, 0.01,
  1.49, 0.14, 0.94, 2.18, 0.05, 0.34, 1.26, 0.29, 2.61,
];

const REQUEST_RETENTION = 0.9;
const MAX_STABILITY = 36500;

function daysBetween(a: Date, b: Date): number {
  return Math.max(0, (a.getTime() - b.getTime()) / (1000 * 60 * 60 * 24));
}

function clampDifficulty(d: number): number {
  return Math.max(1, Math.min(10, d));
}

export function initialFsrs(): FsrsState {
  return {
    stability: 0,
    difficulty: 5,
    reps: 0,
    lapses: 0,
    dueAt: new Date().toISOString(),
  };
}

function initStability(grade: FsrsGrade): number {
  const s = W[grade - 1];
  return Math.max(0.1, s);
}

function initDifficulty(grade: FsrsGrade): number {
  const d = W[4] - (grade - 3) * W[5];
  return clampDifficulty(d);
}

function nextDifficulty(d: number, grade: FsrsGrade): number {
  const deltaD = -W[6] * (grade - 3);
  const newD = d + deltaD * (10 - d) / 9;
  const mean = W[4] * 0.5 + newD * 0.5;
  return clampDifficulty(mean);
}

function retrievability(stability: number, elapsedDays: number): number {
  if (stability <= 0) return 0;
  return Math.pow(1 + elapsedDays / (9 * stability), -1);
}

function nextRecallStability(d: number, s: number, r: number, grade: FsrsGrade): number {
  const hardPenalty = grade === 2 ? W[15] : 1;
  const easyBonus = grade === 4 ? W[16] : 1;
  const factor = Math.exp(W[8]) *
    (11 - d) *
    Math.pow(Math.max(s, 0.01), -W[9]) *
    (Math.exp((1 - r) * W[10]) - 1) *
    hardPenalty * easyBonus;
  return Math.min(MAX_STABILITY, Math.max(0.1, s * (1 + factor)));
}

function nextForgetStability(d: number, s: number, r: number): number {
  const f = W[11] * Math.pow(d, -W[12]) *
    (Math.pow(s + 1, W[13]) - 1) *
    Math.exp((1 - r) * W[14]);
  return Math.min(MAX_STABILITY, Math.max(0.1, f));
}

function intervalForStability(s: number): number {
  if (s <= 0) return 0;
  const i = (s / (9)) * (Math.pow(REQUEST_RETENTION, -1) - 1);
  return Math.max(1, Math.round(i));
}

export function updateFsrs(prev: FsrsState | undefined, grade: FsrsGrade, now: Date = new Date()): FsrsState {
  const state = prev ?? initialFsrs();
  const lastReview = state.lastReviewAt ? new Date(state.lastReviewAt) : null;

  let difficulty: number;
  let stability: number;
  let reps = state.reps + 1;
  let lapses = state.lapses;

  if (state.reps === 0 || state.stability <= 0) {
    difficulty = initDifficulty(grade);
    stability = initStability(grade);
    if (grade === 1) lapses += 1;
  } else {
    const elapsed = lastReview ? daysBetween(now, lastReview) : 0;
    const r = retrievability(state.stability, elapsed);
    difficulty = nextDifficulty(state.difficulty, grade);
    if (grade === 1) {
      stability = nextForgetStability(difficulty, state.stability, r);
      lapses += 1;
    } else {
      stability = nextRecallStability(difficulty, state.stability, r, grade);
    }
  }

  const intervalDays = intervalForStability(stability);
  const due = new Date(now);
  if (grade === 1) {
    due.setMinutes(due.getMinutes() + 10);
  } else {
    due.setDate(due.getDate() + intervalDays);
  }

  return {
    stability,
    difficulty,
    reps,
    lapses,
    lastReviewAt: now.toISOString(),
    dueAt: due.toISOString(),
  };
}

export function getRetrievability(state: FsrsState | undefined, now: Date = new Date()): number {
  if (!state || !state.lastReviewAt || state.stability <= 0) return 0;
  const elapsed = daysBetween(now, new Date(state.lastReviewAt));
  return retrievability(state.stability, elapsed);
}

export function getFsrsPriority(gap: GapItem, now: Date = new Date()): number {
  const state = gap.fsrs;
  if (!state) return gap.reviewCount === 0 ? 0.7 : 0.5;
  const r = getRetrievability(state, now);
  const overdue = Math.max(0, daysBetween(now, new Date(state.dueAt)));
  return (1 - r) + Math.min(overdue / 14, 1) * 0.3;
}

export function mapCorrectnessToGrade(isCorrect: boolean, confidence: 'low' | 'medium' | 'high' = 'medium'): FsrsGrade {
  if (!isCorrect) return 1;
  if (confidence === 'low') return 2;
  if (confidence === 'high') return 4;
  return 3;
}

export function getForgettingCurvePoints(state: FsrsState | undefined, days: number = 30): { day: number; retrievability: number }[] {
  const points: { day: number; retrievability: number }[] = [];
  if (!state || state.stability <= 0) {
    for (let i = 0; i <= days; i += 3) points.push({ day: i, retrievability: 0 });
    return points;
  }
  for (let i = 0; i <= days; i += 3) {
    points.push({ day: i, retrievability: retrievability(state.stability, i) });
  }
  return points;
}

File: ios/FluentFrenchIOS/Models/FSRS.swift
//
//  FSRS.swift
//  FluentFrenchIOS
//
//  Lightweight FSRS-style spaced-repetition scheduler.
//  Mirrors the Expo app's fsrs.ts: each card carries stability, difficulty,
//  and a retrievability curve that drives the optimal next-review date.
//

import Foundation

nonisolated enum ReviewGrade: Int {
    case again = 1
    case hard = 2
    case good = 3
    case easy = 4
}

/// How an answer was produced. The store derives the FSRS grade and the concept
/// evidence weight from this (`Tuning.gradeMapping`, `Tuning.formatEvidenceWeight`)
/// so recognition and production are never scored alike.
nonisolated enum AnswerFormat: String, Codable, CaseIterable {
    case multipleChoice, fillBlank, trueFalse, translation, arrange, match, probe, speaking, converse
}

nonisolated enum FSRS {
    /// Target recall probability when scheduling the next review.
    static let requestRetention: Double = 0.9
    /// Decay constant for the power forgetting curve.
    static let decay: Double = -0.5
    static let factor: Double = 19.0 / 81.0

    static func makeInitialState(grade: ReviewGrade, now: Date = Date()) -> FsrsState {
        let difficulty = initialDifficulty(grade: grade)
        let stability = initialStability(grade: grade)
        let interval = nextInterval(stability: stability)
        return FsrsState(
            stability: stability,
            difficulty: difficulty,
            reps: 1,
            lapses: grade == .again ? 1 : 0,
            lastReviewAt: now,
            dueAt: now.addingTimeInterval(interval * 86_400)
        )
    }

    /// The schedule a gap the learner has NEVER been asked starts from: the `.again`
    /// stability and difficulty (a new item is fragile and due immediately), but with
    /// ZERO reps and ZERO lapses — nobody has answered it, so nobody has missed it.
    /// `lapses` is what the copy layer reads for "you've missed this N×" and for the
    /// lesson headline's "you've slipped on it N times"; seeding it at 1 told a
    /// brand-new learner they had already slipped on every item of their first lesson.
    static func makeUnseenState(now: Date = Date()) -> FsrsState {
        var state = makeInitialState(grade: .again, now: now)
        state.reps = 0
        state.lapses = 0
        state.dueAt = now
        return state
    }

    /// Probability of recall right now given elapsed days since last review.
    static func retrievability(state: FsrsState, now: Date = Date()) -> Double {
        guard let last = state.lastReviewAt else { return 1.0 }
        let elapsedDays = max(0, now.timeIntervalSince(last) / 86_400)
        let base = 1 + factor * (elapsedDays / max(0.1, state.stability))
        return pow(base, decay)
    }

    /// Advance a card's memory state after a review.
    static func review(state: FsrsState?, grade: ReviewGrade, now: Date = Date()) -> FsrsState {
        guard let state else { return makeInitialState(grade: grade, now: now) }

        let r = retrievability(state: state, now: now)
        var difficulty = nextDifficulty(current: state.difficulty, grade: grade)
        difficulty = min(10, max(1, difficulty))

        let newStability: Double
        if grade == .again {
            newStability = lapseStability(difficulty: difficulty, stability: state.stability, retrievability: r)
        } else {
            newStability = successStability(difficulty: difficulty, stability: state.stability, retrievability: r, grade: grade)
        }

        let interval = nextInterval(stability: newStability)
        return FsrsState(
            stability: newStability,
            difficulty: difficulty,
            reps: state.reps + 1,
            lapses: state.lapses + (grade == .again ? 1 : 0),
            lastReviewAt: now,
            dueAt: now.addingTimeInterval(interval * 86_400)
        )
    }

    // MARK: - Internal curve math

    private static func initialDifficulty(grade: ReviewGrade) -> Double {
        let step = Double(grade.rawValue) - Double(ReviewGrade.good.rawValue)
        let d = Tuning.fsrsNeutralDifficulty - step * Tuning.fsrsInitialDifficultyStep
        return min(10, max(1, d))
    }

    private static func initialStability(grade: ReviewGrade) -> Double {
        Tuning.fsrsInitialStability(for: grade)
    }

    /// Difficulty steps away from `.good` by grade; a `.good` answer also drifts
    /// toward `Tuning.fsrsDifficultyMeanReversion`, so a card that keeps being
    /// answered well eases instead of ratcheting up forever on its past lapses.
    private static func nextDifficulty(current: Double, grade: ReviewGrade) -> Double {
        let step = Double(grade.rawValue) - Double(ReviewGrade.good.rawValue)
        var d = current - Tuning.fsrsDifficultyStep * step
        if grade == .good {
            d += (Tuning.fsrsDifficultyMeanReversion - d) * Tuning.fsrsDifficultyReversionRate
        }
        return d
    }

    /// Success growth: stability grows by a term that is larger for easier cards
    /// (headroom below `Tuning.fsrsGrowthDifficultyCeiling`) and for recalls that
    /// were hard-won (low retrievability), scaled by `Tuning.fsrsGrowthRate` and
    /// shaped by the grade (`fsrsHardPenalty` / `fsrsEasyBonus`).
    private static func successStability(difficulty: Double, stability: Double, retrievability: Double, grade: ReviewGrade) -> Double {
        let hardPenalty = grade == .hard ? Tuning.fsrsHardPenalty : 1.0
        let easyBonus = grade == .easy ? Tuning.fsrsEasyBonus : 1.0
        let headroom = exp(Tuning.fsrsGrowthScale * (Tuning.fsrsGrowthDifficultyCeiling - difficulty))
        let recallBoost = 1 + (1 - retrievability) * Tuning.fsrsGrowthRetrievabilityWeight
        let growth = headroom * recallBoost
        return stability * (1 + growth * Tuning.fsrsGrowthRate * hardPenalty * easyBonus)
    }

    /// Multiplicative, monotone lapse: the new stability is a fraction (< 1) of the
    /// old one, so a miss ALWAYS comes back sooner than a hit. Harder cards lose
    /// more; a lapse when recall was already unlikely (low retrievability) is the
    /// expected outcome and keeps more. The ratio is capped below 1 and the result
    /// floored at `Tuning.fsrsMinStability`.
    static func lapseStability(difficulty: Double, stability: Double, retrievability: Double) -> Double {
        let difficultyShape = pow(1 - Tuning.fsrsLapseDifficultyShape, difficulty - Tuning.fsrsNeutralDifficulty)
        let retrievabilityShape = 1 + Tuning.fsrsLapseRetrievabilityShape * (1 - min(1, max(0, retrievability)))
        let ratio = min(Tuning.fsrsLapseMaxRatio, Tuning.fsrsLapseFactorBase * difficultyShape * retrievabilityShape)
        return max(Tuning.fsrsMinStability, stability * ratio)
    }

    private static func nextInterval(stability: Double) -> Double {
        let interval = (stability / factor) * (pow(requestRetention, 1 / decay) - 1)
        return max(0.4, interval)
    }
}
